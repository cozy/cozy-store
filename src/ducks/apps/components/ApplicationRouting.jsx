import React, { Component } from 'react'
import { Route, withRouter } from 'react-router-dom'

import IntentModal from 'cozy-ui/react/IntentModal'
import { translate } from 'cozy-ui/react/I18n'
import Alerter from 'cozy-ui/react/Alerter'

import PermissionsModal from './PermissionsModal'
import UninstallModal from './UninstallModal'
import Install from './Install'
import ApplicationPage from './ApplicationPage'

import { APP_TYPE, REGISTRY_CHANNELS } from 'ducks/apps'

export class ApplicationRouting extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isInstallSuccess: false
    }
    this.getAppFromMatchOrSlug = this.getAppFromMatchOrSlug.bind(this)
  }

  componentWillReceiveProps(nextProps) {
    // don't react on error
    if (nextProps.actionError) return
    // on install success
    if (this.props.isInstalling && !nextProps.isInstalling) {
      const { history, location, parent, t } = this.props
      const pathRegex = new RegExp(`^/${parent}/([^/]*)/.*`)
      const matches = location.pathname.match(pathRegex)
      if (!matches || matches.length < 1) return history.replace(`/${parent}/`)
      const app = this.getAppFromMatchOrSlug(null, matches[1])
      if (app.type === APP_TYPE.KONNECTOR) {
        return history.replace(`/${parent}/${app.slug}/configure`)
      } else {
        Alerter.success(t('app_modal.install.message.install_success'), {
          duration: 3000
        })
        return history.replace(`/${parent}/${app.slug}`)
      }
    } else if (this.props.isUninstalling && !nextProps.isUninstalling) {
      const { history, location, parent, t } = this.props
      const pathRegex = new RegExp(`^/${parent}/([^/]*)/.*`)
      const matches = location.pathname.match(pathRegex)
      if (!matches || matches.length < 1) return history.replace(`/${parent}/`)
      const app = this.getAppFromMatchOrSlug(null, matches[1])
      Alerter.success(t('app_modal.uninstall.message.success'), {
        duration: 3000
      })
      return history.replace(`/${parent}/${app.slug}`)
    }
  }

  getAppFromMatchOrSlug(match, slug) {
    const appsArray = this.props.apps || this.props.installedApps || []
    const appSlug = slug || (match && match.params && match.params.appSlug)
    if (!appsArray.length || !appSlug) return null
    const app = appsArray.find(app => app.slug === appSlug)
    return app
  }

  render() {
    const {
      fetchLatestApp,
      installApp,
      uninstallApp,
      updateApp,
      isFetching,
      isAppFetching,
      isInstalling,
      parent,
      history,
      actionError,
      fetchError
    } = this.props
    return (
      <div>
        <Route
          path={`/${parent}/:appSlug`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            return <ApplicationPage app={app} parent={parent} />
          }}
        />
        <Route
          path={`/${parent}/:appSlug/channel/:channel`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app || !app.isInRegistry) return history.replace(`/${parent}`)
            const channel = match.params.channel
            const isChannelAvailable = Object.values(
              REGISTRY_CHANNELS
            ).includes(channel)
            if (!isChannelAvailable) {
              return history.replace(`/${parent}/${app.slug}/manage`)
            }
            return (
              <Install
                installApp={app.installed ? updateApp : installApp}
                parent={`/${parent}`}
                fetchApp={chan => fetchLatestApp(app.slug, chan)}
                isAppFetching={isAppFetching}
                installError={actionError}
                fetchError={fetchError}
                app={app}
                isInstalling={isInstalling}
                channel={channel}
              />
            )
          }}
        />
        <Route
          path={`/${parent}/:appSlug/manage`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            if (app && app.installed) {
              if (!app.uninstallable)
                return history.replace(`/${parent}/${app.slug}`)
              return (
                <UninstallModal
                  uninstallApp={uninstallApp}
                  parent={`/${parent}`}
                  uninstallError={actionError}
                  app={app}
                />
              )
            } else {
              return (
                <Install
                  installApp={installApp}
                  parent={`/${parent}`}
                  installError={actionError}
                  app={app}
                  isInstalling={isInstalling}
                />
              )
            }
          }}
        />
        <Route
          path={`/${parent}/:appSlug/permissions`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            return <PermissionsModal app={app} parent={`/${parent}`} />
          }}
        />
        <Route
          path={`/${parent}/:appSlug/configure`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            const goToApp = () => history.replace(`/${parent}/${app.slug}`)
            if (app && app.installed && app.type === APP_TYPE.KONNECTOR) {
              return (
                <IntentModal
                  action="CREATE"
                  doctype="io.cozy.accounts"
                  options={{ slug: app.slug }}
                  dismissAction={goToApp}
                  onComplete={goToApp}
                  mobileFullscreen
                  overflowHidden
                  size="small"
                  height="35rem"
                />
              )
            } else {
              return goToApp()
            }
          }}
        />
      </div>
    )
  }
}

export default withRouter(translate()(ApplicationRouting))
