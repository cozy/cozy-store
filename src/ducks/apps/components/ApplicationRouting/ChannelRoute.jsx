import { REGISTRY_CHANNELS } from 'ducks/apps'
import ChannelModal from 'ducks/apps/components/ChannelModal'
import React from 'react'
import { useParams } from 'react-router-dom'

export const ChannelRoute = ({ getApp, isFetching, parent, redirectTo }) => {
  const params = useParams()
  if (isFetching) return null
  const app = getApp(params)
  const appPath = `/${parent}/${(app && app.slug) || ''}`
  if (!app) return redirectTo(`/${parent}`)
  const channel = params.channel
  const isChannelAvailable = Object.values(REGISTRY_CHANNELS).includes(channel)
  if (!isChannelAvailable) {
    return redirectTo(appPath)
  }
  return (
    <ChannelModal
      appSlug={app.slug}
      channel={channel}
      dismissAction={() => redirectTo(appPath)}
      onCurrentChannel={() => redirectTo(appPath)}
      onNotHandled={() => redirectTo(appPath)}
      onSuccess={() => redirectTo(appPath)}
    />
  )
}

export default ChannelRoute
