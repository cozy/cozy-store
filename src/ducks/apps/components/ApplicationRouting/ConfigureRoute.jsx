import React from 'react'
import { Route, useParams } from 'react-router-dom'

import ConfigureModal from 'ducks/apps/components/ConfigureModal'

export const ConfigureRoute = ({ getApp, isFetching, parent, redirectTo }) => {
  const params = useParams()
  if (isFetching) return null
  const app = getApp(params)
  if (!app) return redirectTo(`/${parent}`)
  const appPath = `/${parent}/${app.slug}`
  const redirectToApp = () => redirectTo(appPath)
  return (
    <ConfigureModal
      appSlug={app.slug}
      dismissAction={redirectToApp}
      onNotInstalled={redirectToApp}
      onSuccess={redirectToApp}
      onWebApp={redirectToApp}
    />
  )

  // <Route
  //   path={`/${parent}/:appSlug/configure`}
  //   render={({ match }) => {
  //     if (isFetching) return null
  //     const app = getApp(match)
  //     if (!app) return redirectTo(`/${parent}`)
  //     const appPath = `/${parent}/${app.slug}`
  //     const redirectToApp = () => redirectTo(appPath)
  //     return (
  //       <ConfigureModal
  //         appSlug={app.slug}
  //         dismissAction={redirectToApp}
  //         onNotInstalled={redirectToApp}
  //         onSuccess={redirectToApp}
  //         onWebApp={redirectToApp}
  //       />
  //     )
  //   }}
  // />
}

export default ConfigureRoute
