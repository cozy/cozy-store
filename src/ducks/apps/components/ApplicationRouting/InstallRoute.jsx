import React from 'react'
import { Route } from 'react-router-dom'

import InstallModal from 'ducks/apps/components/InstallModal'

export const InstallRoute = ({ getApp, isFetching, parent, redirectTo }) => (
  <Route
    path={`/${parent}/:appSlug/install`}
    render={({ match }) => {
      if (isFetching) return null
      const app = getApp(match)

      if (!app) {
        return redirectTo(`/${parent}`)
      }

      const appPath = `/${parent}/${(app && app.slug) || ''}`
      const configurePath = `${appPath}/configure`
      const redirectToApp = () => redirectTo(appPath)
      const redirectToConfigure = () => redirectTo(configurePath)
      return (
        <InstallModal
          app={app}
          onInstalled={redirectToApp}
          dismissAction={redirectToApp}
          redirectToConfigure={redirectToConfigure}
          redirectToApp={redirectToApp}
        />
      )
    }}
  />
)

export default InstallRoute
