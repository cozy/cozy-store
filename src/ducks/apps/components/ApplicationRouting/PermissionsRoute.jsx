import PermissionsModal from '@/ducks/apps/components/PermissionsModal'
import IntentPermissionsModal from '@/ducks/components/intents/IntentPermissionsModal'
import React from 'react'
import { useLocation, useParams } from 'react-router-dom'

export const PermissionsRoute = ({
  getApp,
  isFetching,
  parent,
  redirectTo,
  intentData
}) => {
  const params = useParams()
  const { search } = useLocation()

  if (isFetching) return null

  const app = getApp(params)

  if (!app) return redirectTo(`/${parent}${search}`)

  const Modal = intentData ? IntentPermissionsModal : PermissionsModal

  return <Modal app={app} parent={`/${parent}`} intentData={intentData} />
}

export default PermissionsRoute
