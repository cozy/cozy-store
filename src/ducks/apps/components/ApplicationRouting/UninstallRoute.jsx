import React from 'react'
import { useParams } from 'react-router-dom'

import UninstallModal from 'ducks/apps/components/UninstallModal'

export const UninstallRoute = ({ getApp, isFetching, parent, redirectTo }) => {
  const params = useParams()

  if (isFetching) return null

  const parentPath = `/${parent}`
  const app = getApp(params)
  if (!app) return redirectTo(parentPath)

  const appPath = `${parentPath}/${app.slug}`
  return (
    <UninstallModal
      appSlug={app.slug}
      dismissAction={() => redirectTo(appPath)}
      onNotInstalled={() => redirectTo(appPath)}
      onSuccess={() => redirectTo(appPath)}
    />
  )
}

export default UninstallRoute
