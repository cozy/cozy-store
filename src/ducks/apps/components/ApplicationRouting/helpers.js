import { isFlagshipApp } from 'cozy-device-helper'

export const redirectToConfigure = async ({
  intentData,
  compose,
  app,
  parent,
  redirectTo,
  onTerminate
}) => {
  const { data } = intentData || {}
  const enableConfiguration =
    data && (typeof data.configure === 'undefined' || data.configure)

  const appPath = `/${parent}/${app?.slug || ''}`
  const configurePath = `${appPath}/configure`

  const redirectToApp = () => redirectTo(appPath)

  if (intentData) {
    if (enableConfiguration) {
      /*
        Within an Intent, we cannot use the `/configure` route, because it ultimately returns the `IntentIframe` component which creates a new `iframe` inside the first one.
        This results in a `CSP:frame-ancestors` problem
        It is therefore necessary to favor the use of the `compose` service of `cozy-inter-app` which will add the new `iframe` next to the first
      */
      await compose(
        'CREATE',
        'io.cozy.accounts',
        { slug: app.slug }
      )
    } else {
      onTerminate(app)
    }
    return
  }

  if (isFlagshipApp()) {
    redirectToApp()
  } else {
    redirectTo(configurePath)
  }
}
