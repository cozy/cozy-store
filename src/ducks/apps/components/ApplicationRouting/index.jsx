import ApplicationPage from 'ducks/apps/components/ApplicationPage'
import ChannelRoute from 'ducks/apps/components/ApplicationRouting/ChannelRoute'
import ConfigureRoute from 'ducks/apps/components/ApplicationRouting/ConfigureRoute'
import InstallRoute from 'ducks/apps/components/ApplicationRouting/InstallRoute'
import PermissionsRoute from 'ducks/apps/components/ApplicationRouting/PermissionsRoute'
import UninstallRoute from 'ducks/apps/components/ApplicationRouting/UninstallRoute'
import React, { Component } from 'react'
import {
  Outlet,
  Route,
  Routes,
  useLocation,
  useNavigate
} from 'react-router-dom'

import { translate } from 'cozy-ui/transpiled/react/I18n'

const OutletWrapper = ({ Component }) => (
  <>
    <Component />
    <Outlet />
  </>
)

export class ApplicationRouting extends Component {
  mainPage = React.createRef()

  getAppFromMatchOrSlug = (params, slug) => {
    const appsArray = this.props.apps || this.props.installedApps || []
    const appSlug = slug || (params && params.appSlug)
    if (!appsArray.length || !appSlug) return null
    const app = appsArray.find(app => app.slug === appSlug)
    return app
  }

  redirectTo = target => {
    const { navigate, location } = this.props
    navigate(target + location.search, { replace: true })
    return null
  }

  render() {
    const { isFetching, parent } = this.props
    return (
      <div className="sto-modal-page" ref={this.mainPage}>
        <Routes>
          <Route
            path=":appSlug"
            element={
              <OutletWrapper
                Component={() => (
                  <ApplicationPage
                    parent={parent}
                    isFetching={isFetching}
                    getApp={this.getAppFromMatchOrSlug}
                    redirectTo={this.redirectTo}
                    mainPageRef={this.mainPage}
                  />
                )}
              />
            }
          >
            <Route
              path="channel/:channel"
              element={
                <ChannelRoute
                  getApp={this.getAppFromMatchOrSlug}
                  isFetching={isFetching}
                  parent={parent}
                  redirectTo={this.redirectTo}
                />
              }
            />
            <Route
              path="install"
              element={
                <InstallRoute
                  getApp={this.getAppFromMatchOrSlug}
                  isFetching={isFetching}
                  parent={parent}
                  redirectTo={this.redirectTo}
                />
              }
            />
            <Route
              path="uninstall"
              element={
                <UninstallRoute
                  getApp={this.getAppFromMatchOrSlug}
                  isFetching={isFetching}
                  parent={parent}
                  redirectTo={this.redirectTo}
                />
              }
            />
            <Route
              path="permissions"
              element={
                <PermissionsRoute
                  getApp={this.getAppFromMatchOrSlug}
                  isFetching={isFetching}
                  parent={parent}
                  redirectTo={this.redirectTo}
                />
              }
            />
            <Route
              path="configure"
              element={
                <ConfigureRoute
                  getApp={this.getAppFromMatchOrSlug}
                  isFetching={isFetching}
                  parent={parent}
                  redirectTo={this.redirectTo}
                />
              }
            />
          </Route>
        </Routes>
      </div>
    )
  }
}

const ApplicationRoutingWrapper = props => {
  const navigate = useNavigate()
  const location = useLocation()
  return (
    <ApplicationRouting {...props} navigate={navigate} location={location} />
  )
}

export default translate()(ApplicationRoutingWrapper)
