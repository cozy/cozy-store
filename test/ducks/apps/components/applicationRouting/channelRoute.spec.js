'use strict'

/* eslint-env jest */

import { ChannelRoute } from '@/ducks/apps/components/ApplicationRouting/ChannelRoute'
import { shallow } from 'enzyme'
import React from 'react'
import { useParams } from 'react-router-dom'

import mockApps from '../../_mockApps'

const getAppMock = params => mockApps.find(a => a.slug === params.appSlug)

const getProps = (isFetching = false, getApp = getAppMock) => ({
  fetchLatestApp: jest.fn(),
  isFetching,
  parent: 'myapps',
  getApp,
  redirectTo: jest.fn(() => null)
})

jest.mock('react-router-dom', () => ({
  useParams: jest.fn(),
  useLocation: jest.fn()
}))

describe('ChannelRoute component', () => {
  it('should display install modal if app found and in the registry and installed', () => {
    // collect in mockApps is installed and isInRegistry
    useParams.mockReturnValue({ appSlug: 'collect', channel: 'beta' })
    const props = getProps()
    const component = shallow(<ChannelRoute {...props} />)
    expect(component).toMatchSnapshot()
  })

  it('should display install modal if app found and in the registry even if not installed', () => {
    // photos in mockApps is isInRegistry
    useParams.mockReturnValue({ appSlug: 'photos', channel: 'beta' })
    const props = getProps()
    const component = shallow(<ChannelRoute {...props} />)
    expect(component).toMatchSnapshot()
  })

  it('should display nothing if isFetching', () => {
    // collect in mockApps is installed and isInRegistry
    useParams.mockReturnValue({ appSlug: 'collect' })
    const props = getProps(true)
    const component = shallow(<ChannelRoute {...props} />)
    expect(component.isEmptyRender()).toBe(true)
  })

  it('should redirectTo parent if no app found', () => {
    useParams.mockReturnValue({})
    const props = getProps(false, jest.fn())
    const component = shallow(<ChannelRoute {...props} />)
    expect(props.redirectTo.mock.calls.length).toBe(1)
    expect(props.redirectTo.mock.calls[0][0]).toBe(`/${props.parent}`)
    expect(component.isEmptyRender()).toBe(true)
  })

  it('should redirectTo app page if app channel not available', () => {
    // collect in mockApps is installed and isInRegistry
    const params = { appSlug: 'collect', channel: 'jeanluc' }
    useParams.mockReturnValue(params)
    const props = getProps()
    const component = shallow(<ChannelRoute {...props} />)
    expect(props.redirectTo.mock.calls.length).toBe(1)
    expect(props.redirectTo.mock.calls[0][0]).toBe(
      `/${props.parent}/${params.appSlug}`
    )
    expect(component.isEmptyRender()).toBe(true)
  })
})
