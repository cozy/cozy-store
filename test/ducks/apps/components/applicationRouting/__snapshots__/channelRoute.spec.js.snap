// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelRoute component should display install modal if app found and in the registry and installed 1`] = `
<withClient(Wrapper)
  appSlug="collect"
  channel="beta"
  dismissAction={[Function]}
  onCurrentChannel={[Function]}
  onNotHandled={[Function]}
  onSuccess={[Function]}
/>
`;

exports[`ChannelRoute component should display install modal if app found and in the registry even if not installed 1`] = `
<withClient(Wrapper)
  appSlug="photos"
  channel="beta"
  dismissAction={[Function]}
  onCurrentChannel={[Function]}
  onNotHandled={[Function]}
  onSuccess={[Function]}
/>
`;
